nx/lib/eigh_block.ex

```diff
@@ -0,0 +1,254 @@
+  which is built on the approach in:
+  Brent, R. P., & Luk, F. T. (1985). The solution of singular-value
+  and symmetric eigenvalue problems on multiprocessor arrays.
+  SIAM Journal on Computing, 6(1), 69-84. https://doi.org/10.1137/0906007
+  """
+  require Nx
+
+  import Nx.Defn
+
+  defn calc_rot(tl, tr, br) do
+    a = Nx.take_diagonal(br)
+    b = Nx.take_diagonal(tr)
+    c = Nx.take_diagonal(tl)
+
+    tau = (a - c) / (2 * b)
+    t = Nx.sqrt(1 + Nx.pow(tau, 2))
+    t = Nx.select(Nx.greater_equal(tau, 0), 1 / (tau + t), 1 / (tau - t))
+
+    pred = Nx.less_equal(Nx.abs(b), 0.1 * 1.0e-4 * Nx.min(Nx.abs(a), Nx.abs(c)))
+    t = Nx.select(pred, 0.0, t)
+
+    c = 1.0 / Nx.sqrt(1.0 + Nx.pow(t, 2))
+    s = t * c
+
+    rt1 = tl - t * tr
+    rt2 = br + t * tr
+
+    {rt1, rt2, c, s}
+  end
+
+  defn sq_norm(tl, tr, bl, br) do
+    Nx.sum(Nx.pow(tl, 2) + Nx.pow(tr, 2) + Nx.pow(bl, 2) + Nx.pow(br, 2))
+  end
+
+  defn off_norm(tl, tr, bl, br) do
+    {n, _} = Nx.shape(tl)
+    diag = Nx.broadcast(0, {n})
+    o_tl = Nx.put_diagonal(tl, diag)
+    o_br = Nx.put_diagonal(br, diag)
+
+    Nx.sum(Nx.pow(o_tl, 2) + Nx.pow(tr, 2) + Nx.pow(bl, 2) + Nx.pow(o_br, 2))
+  end
+
+  @doc """
+  Calculates the Frobenius norm and the norm of the off-diagonals from
+  the submatrices. Used to calculate convergeance.
+  """
+  defn norms(tl, tr, bl, br) do
+    frob = sq_norm(tl, tr, bl, br)
+    off = off_norm(tl, tr, bl, br)
+    {frob, off}
+  end
+
+  defn eigh(matrix) do
+    matrix
+    |> Nx.revectorize([collapsed_axes: :auto],
+      target_shape: {Nx.axis_size(matrix, -2), Nx.axis_size(matrix, -1)}
+    )
+    |> decompose()
+    |> then(fn {w, v} ->
+      revectorize_result({w, v}, matrix)
+    end)
+  end
+
+  deftransformp revectorize_result({eigenvals, eigenvecs}, a) do
+    shape = Nx.shape(a)
+
+    {
+      Nx.revectorize(eigenvals, a.vectorized_axes,
+        target_shape: Tuple.delete_at(shape, tuple_size(shape) - 1)
+      ),
+      Nx.revectorize(eigenvecs, a.vectorized_axes, target_shape: shape)
+    }
+  end
+
+  defn decompose(matrix) do
+    {n, _} = Nx.shape(matrix)
+
+    if n > 1 do
+      m_decompose(matrix)
+    else
+      {Nx.tensor([1], type: matrix.type), Nx.take_diagonal(matrix)}
+    end
+  end
+
+  defn m_decompose(matrix) do
+    {n, _} = Nx.shape(matrix)
+    i_n = n - 1
+    {mid, _} = Nx.shape(matrix[[0..i_n//2, 0..i_n//2]])
+    i_mid = mid - 1
+
+    {tl, tr, bl, br} =
+      {matrix[[0..i_mid, 0..i_mid]], matrix[[0..i_mid, mid..i_n]], matrix[[mid..i_n, 0..i_mid]],
+       matrix[[mid..i_n, mid..i_n]]}
+
+    # Pad if not even
+    {tl, tr, bl, br} =
+      if Nx.remainder(n, 2) == 1 do
+        tr = Nx.pad(tr, 0, [{0, 0, 0}, {0, 1, 0}])
+        bl = Nx.pad(bl, 0, [{0, 1, 0}, {0, 0, 0}])
+        br = Nx.pad(br, 0, [{0, 1, 0}, {0, 1, 0}])
+        {tl, tr, bl, br}
+      else
+        {tl, tr, bl, br}
+      end
+
+    # Initialze tensors to hold eigenvectors
+    v_tl = Nx.eye(mid, type: :f32)
```

Comment:
why force `f32` here? Is this another case where the algorithm just fails on f64?
Perhaps this should be masked underneath the implementation if it's the case.